File: src/main/java/ru/ioleksiv/telegram/bot/api/annotations/filter/secondary/UserFilter.java
package ru.ioleksiv.telegram.bot.api.annotations.filter.secondary;

import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.BooleanFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.StringFilter;
import ru.ioleksiv.telegram.bot.api.model.annotation.AnnotationState;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Documented
public @interface UserFilter {

    AnnotationState value() default AnnotationState.ON;

    StringFilter firstName() default @StringFilter(AnnotationState.OFF);

    StringFilter lastName() default @StringFilter(AnnotationState.OFF);

    StringFilter username() default @StringFilter(AnnotationState.OFF);

    StringFilter languageCode() default @StringFilter(AnnotationState.OFF);

    BooleanFilter bot() default @BooleanFilter(AnnotationState.OFF);

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/api/annotations/filter/composite/UserFilter.kt
package io.github.oleksivio.telegram.bot.api.annotations.filter.composite

import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.BooleanFilter
import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.StringFilter
import io.github.oleksivio.telegram.bot.api.model.annotation.AnnotationState

@Retention
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@MustBeDocumented
annotation class UserFilter(
        val status: AnnotationState = AnnotationState.ON,
        val validator: Array<String> = [],
        val firstName: StringFilter = StringFilter(status = AnnotationState.OFF),
        val lastName: StringFilter = StringFilter(status = AnnotationState.OFF),
        val username: StringFilter = StringFilter(status = AnnotationState.OFF),
        val languageCode: StringFilter = StringFilter(status = AnnotationState.OFF),
        val bot: BooleanFilter = BooleanFilter(status = AnnotationState.OFF)
)
